waku/v2/node/wakunode2.go

```diff
@@ -617,21 +619,32 @@ func (w *WakuNode) startKeepAlive(t time.Duration) {
 	}()
 }
 
-func pingPeer(ctx context.Context, wg *sync.WaitGroup, host host.Host, peer peer.ID) {
-	wg.Add(1)
-	defer wg.Done()
+func (w *WakuNode) pingPeer(peer peer.ID) {
+	w.keepAliveMutex.Lock()
+	defer w.keepAliveMutex.Unlock()
+	defer w.wg.Done()
 
-	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	ctx, cancel := context.WithTimeout(w.ctx, 3*time.Second)
 	defer cancel()
 
 	log.Debug("Pinging ", peer)
-	pr := ping.Ping(ctx, host, peer)
+	pr := ping.Ping(ctx, w.host, peer)
 	select {
 	case res := <-pr:
 		if res.Error != nil {
+			w.keepAliveFails[peer]++
 			log.Debug(fmt.Sprintf("Could not ping %s: %s", peer, res.Error.Error()))
+		} else {
+			w.keepAliveFails[peer] = 0
 		}
 	case <-ctx.Done():
+		w.keepAliveFails[peer]++
 		log.Debug(fmt.Sprintf("Could not ping %s: %s", peer, ctx.Err()))
 	}
+
+	if w.keepAliveFails[peer] > 2 && w.host.Network().Connectedness(peer) == network.Connected {
```

Comment:
nit: make the 2 a constant with a comment for easy future tweak ;)

Reply:
cool!